### PR TITLE
feat: implement SubmitConsent tx

### LINKED ISSUE #513: Implement SubmitConsent tx

### PR DESCRIPTION
**Pull request by gyuguen**
Close #513 

### Implement
* Added submitConsent
  * verify signature
  * verify signed oracle
  * verify deal
  * verify duplicate certificate
  * send reward
  * post process of deal

### CODE REVIEW

**Comment by gyuguen** [comment]
I will add comment soon :)

File: app/app.go [hunk 1]
```diff
@@ -428,12 +431,20 @@ func New(
 		app.BankKeeper,
 	)
 
+	app.oracleKeeper = *oraclekeeper.NewKeeper(
```
Comment (gyuguen) [inline app/app.go:434]: This oracle setup is already set in another PR.
I will change this code when merging

File: x/datadeal/keeper/certificate.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	pricePerData := deal.GetPricePerData()
+
+	dealBalance := k.bankKeeper.GetBalance(ctx, dealAccAddr, assets.MicroMedDenom)
+	if dealBalance.IsLT(sdk.NewCoin(assets.MicroMedDenom, pricePerData.TruncateInt())) {
+		return fmt.Errorf("not enough balance in deal")
+	}
+
+	// TODO calculate oracle commission
```
Comment (gyuguen) [inline x/datadeal/keeper/certificate.go:78]: When after merged oracle feature, I will add a oracle commission.

**Comment by gyuguen** [comment]
I added `sdk.Msg` in `message_deal.go`.
Because we are getting errors.
```
$ panacead q deal
        /home/junney/workspace/panacea-core/x/datadeal/module.go:55
        /home/junney/workspace/panacea-core/x/datadeal/module.go:55
github.com/cosmos/cosmos-sdk/types/module.BasicManager.RegisterInterfaces(...)
        /home/junney/go/pkg/mod/github.com/medibloc/cosmos-sdk@v0.45.9-panacea.1/types/module/module.go:86
github.com/medibloc/panacea-core/v2/app.MakeEncodingConfig()

...
```

File: x/oracle/keeper/keeper.go [hunk 3]
```diff
@@ -35,3 +38,40 @@ func NewKeeper(
 		paramSpace: paramSpace,
 	}
 }
+
+func (k Keeper) VerifySignature(ctx sdk.Context, msg codec.ProtoMarshaler, sigBz []byte) error {
```
Comment (H4NLee) [inline x/oracle/keeper/keeper.go:42]: Why don't we name this function more specific? e.g. `VerifyOracleSignature`.
It seems too general to me.

File: x/oracle/keeper/keeper.go [hunk 4]
```diff
@@ -35,3 +38,40 @@ func NewKeeper(
...
+	if err != nil {
+		return err
+	}
+
+	// TODO Check is oracle registered?
+
+	// TODO Check is registered oracle's uniqueId correct?
+
+	return nil
+}
```
Comment (H4NLee) [inline x/oracle/keeper/keeper.go:77]: Both TODOs make sense to me.
If there are some critical bugs in old version oracle, then only certificates issued by the upgraded version of oracle should be valid. That is, unique ID should also be added to certificate. wdyt?

If then, this function probably should check if the oracle is registered or not. The other one is validation of certificate.

File: x/datadeal/types/message_deal.go [hunk 5]
```diff
@@ -16,27 +16,20 @@ func (m *MsgCreateDeal) Type() string {
...
 		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "no data schema")
 	}
-
-	budget := m.Budget
-	if budget == nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "budget is empty")
+	if m.MaxNumData <= 0 {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "max num of data is negative number")
+	}
+	if m.Budget == nil {
```
Comment (H4NLee) [inline x/datadeal/types/message_deal.go:25]: It seems that a negative check of the budget should also be done.

File: x/datadeal/keeper/certificate.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	existUnsignedCert, err := k.GetCertificate(ctx, dealID, dataHash)
+	if err != types.ErrCertificateNotFound {
+		if existUnsignedCert != nil {
+			return fmt.Errorf("already exist certificate. dataHash: %s", dataHash)
+		} else {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (H4NLee) [inline x/datadeal/keeper/certificate.go:58]: I think we can simply return if the store has the certificate or not.
```go
func (k Keeper) isProvidedCert(ctx sdk.Context, dealID uint64, dataHash string) bool {
    store := ctx.KVStore(k.storeKey)
    return store.Has(types.GetCertificateKey(dealID, dataHash))
}
```

File: x/oracle/keeper/keeper.go (hunk 4 above)
Comment (gyuguen) [reply x/oracle/keeper/keeper.go:77]: Umm, I think this function is only checks Oracle.

As you said, It seems that we need to verify the uniqueID of the certificate.
I will implement a function to verify the uniqueID of a certificate.

File: x/datadeal/types/message_deal.go (hunk 5 above)
Comment (gyuguen) [reply x/datadeal/types/message_deal.go:25]: It was already included in `m.Budget.IsValid()` :)

File: x/datadeal/keeper/certificate.go (hunk 6 above)
Comment (gyuguen) [reply x/datadeal/keeper/certificate.go:58]: It seems good. I will fix soon.

Comment (gyuguen) [reply x/datadeal/keeper/certificate.go]: fixed: https://github.com/medibloc/panacea-core/pull/533/commits/fec02e4707f772c7ada78e92414e0aa8d2ea5a30

File: x/oracle/keeper/keeper.go (hunk 3 above)
Comment (gyuguen) [reply x/oracle/keeper/keeper.go]: fixed: https://github.com/medibloc/panacea-core/pull/533/commits/fec02e4707f772c7ada78e92414e0aa8d2ea5a30
Thank you.

**Comment by youngjoon-lee** [comment]
> I added `sdk.Msg` in `message_deal.go`. (https://github.com/medibloc/panacea-core/pull/533/commits/26101ad563b4c31ce4c5b1fc132231ee0a2ce657)

Could you share what the actual error message was? I can see only call stacks from your comment.

**Review by audtlr24 — APPROVED** [review]
lgtm! Thanks

**Review by H4NLee — APPROVED** [review]
LGTM

**Comment by gyuguen** [comment]
> > I added `sdk.Msg` in `message_deal.go`. ([26101ad](https://github.com/medibloc/panacea-core/commit/26101ad563b4c31ce4c5b1fc132231ee0a2ce657))
> 
> Could you share what the actual error message was? I can see only call stacks from your comment.

It seems to happen because of the line below.
https://github.com/medibloc/panacea-core/blob/ft/513/implement_submit_consent/x/datadeal/types/codec.go#L19

I wrote it ahead of time because we are gonna have to implement this struct anyway.

**Comment by gyuguen** [comment]
> > I added `sdk.Msg` in `message_deal.go`. ([26101ad](https://github.com/medibloc/panacea-core/commit/26101ad563b4c31ce4c5b1fc132231ee0a2ce657))
> 
> Could you share what the actual error message was? I can see only call stacks from your comment.

In fact, this error occurs from the init command.

```
$ panacead init
github.com/medibloc/panacea-core/v2/x/datadeal.AppModuleBasic.RegisterInterfaces(...)
        /home/junney/workspace/panacea-core/x/datadeal/module.go:55
github.com/cosmos/cosmos-sdk/types/module.BasicManager.RegisterInterfaces(...)
        /home/junney/go/pkg/mod/github.com/medibloc/cosmos-sdk@v0.45.9-panacea.1/types/module/module.go:86
github.com/medibloc/panacea-core/v2/app.MakeEncodingConfig()
        /home/junney/workspace/panacea-core/app/encoding.go:14 +0x1ca
github.com/medibloc/panacea-core/v2/cmd/panacead/cmd.NewRootCmd()
        /home/junney/workspace/panacea-core/cmd/panacead/cmd/root.go:52 +0x5f
main.main()
        /home/junney/workspace/panacea-core/cmd/panacead/main.go:12 +0x19

```

**Review by youngjoon-lee — APPROVED** [review]
LGTM. Please rename the title of this PR as below, when you merge this PR to the main.
```
feat: implement SubmitConsent tx
```

**Pull request merged**